File: numberToPerfectoid2.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>

std::map<int,int> prime_factors(int n) {
    std::map<int,int> factors;
    for (int p = 2; p*p <= n; ++p) {
        while (n % p == 0) {
            factors[p]++;
            n /= p;
        }
    }
    if (n > 1) factors[n] = 1;
    return factors;
}

int gcd(int a, int b) {
    while (b != 0) {
        int t = b;
        b = a % b;
        a = t;
    }
    return a;
}

int lcm(int a, int b) {
    return a / gcd(a,b) * b;
}

std::map<int,int> min_exponents(const std::vector<std::map<int,int>>& factor_lists, const std::vector<int>& all_primes) {
    std::map<int,int> min_exp;
    for (int p : all_primes) {
        int m = INT_MAX;
        for (auto &f : factor_lists)
            m = std::min(m, f.count(p) ? f.at(p) : 0);
        min_exp[p] = m;
    }
    return min_exp;
}

std::map<int,int> max_exponents(const std::vector<std::map<int,int>>& factor_lists, const std::vector<int>& all_primes) {
    std::map<int,int> max_exp;
    for (int p : all_primes) {
        int M = 0;
        for (auto &f : factor_lists)
            M = std::max(M, f.count(p) ? f.at(p) : 0);
        max_exp[p] = M;
    }
    return max_exp;
}

int main() {
    std::vector<int> numbers = {12, 18, 30};
    std::vector<std::map<int,int>> factor_lists;
    std::map<int,bool> prime_set;

    // Primfaktorzerlegung
    for (int n : numbers) {
        auto f = prime_factors(n);
        factor_lists.push_back(f);
        for (auto &p : f) prime_set[p.first] = true;
    }

    // Alle Primzahlen
    std::vector<int> all_primes;
    for (auto &p : prime_set) all_primes.push_back(p.first);

    auto min_exp = min_exponents(factor_lists, all_primes);
    auto max_exp = max_exponents(factor_lists, all_primes);

    // ggT und kgV berechnen
    int ggT = 1, kgV = 1;
    for (int p : all_primes) {
        ggT *= pow(p, min_exp[p]);
        kgV *= pow(p, max_exp[p]);
    }

    // Perfektoide Zahl in Produktform ausgeben
    std::cout << "ggT: " << ggT << ", kgV: " << kgV << "\n";
    std::cout << "Perfektoide Zahl (Produktform):\n";
    for (int p : all_primes)
        std::cout << p << "^" << min_exp[p] << ".." << p << "^" << max_exp[p] << "\n";
}

File: docs/cppPlusPy/numberToPerfectoid2.py
from math import gcd
from functools import reduce
import sympy

def prime_factors(n):
    """Primfaktorzerlegung als Dictionary {p: exponent}"""
    return sympy.factorint(n)

def lcm(a, b):
    return abs(a*b) // gcd(a, b)

def gcd_list(nums):
    return reduce(gcd, nums)

def lcm_list(nums):
    return reduce(lcm, nums)

def perfeqize_product(numbers):
    """Erzeugt eine 'perfektoide Zahl' aus ganzen Zahlen"""
    factor_lists = [prime_factors(n) for n in numbers]

    # Alle Primzahlen sammeln
    all_primes = set()
    for f in factor_lists:
        all_primes.update(f.keys())

    # ggT und kgV Exponenten pro Primzahl
    min_exp = {}
    max_exp = {}
    for p in all_primes:
        exps = [f.get(p, 0) for f in factor_lists]
        min_exp[p] = min(exps)
        max_exp[p] = max(exps)

    # Fertige p-adische/perfektoide Zahl als Produktform
    # Beispiel: 2^min 3^min ... bis 2^max 3^max ...
    # Wir erzeugen ein Dictionary: {"ggT": ..., "kgV": ...}
    ggT = 1
    kgV = 1
    for p in all_primes:
        ggT *= p ** min_exp[p]
        kgV *= p ** max_exp[p]

    # Perfektoide Zahl symbolisch als Projektivlimit
    perfektoid = {p: f"{p}^{min_exp[p]}..{p}^{max_exp[p]}" for p in all_primes}

    return {
        "numbers": numbers,
        "ggT": ggT,
        "kgV": kgV,
        "perfektoid_product_form": perfektoid
    }

# Beispiel
numbers = [12, 18, 30]
perfektoid = perfeqize_product(numbers)
from pprint import pprint
pprint(perfektoid)
